File: Search/src/main/java/bfs/Solution200.java
package bfs;

import java.util.ArrayDeque;
import java.util.Queue;

public class Solution200 {
    private int m, n;
    private boolean[][] visited;
    private char[][] grid;

    public int numIslands(char[][] grid) {
        this.grid = grid;
        m = grid.length;
        n = grid[0].length;
        visited = new boolean[m][n];

        int ans = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    ans++;
                    bfs(i, j);
                }
            }
        }
        return ans;

    }

    private void bfs(int i, int j) {
        int[] dx = { 1, 0, 0, -1 };
        int[] dy = { 0, 1, -1, 0 };
        Queue<int[]> q = new ArrayDeque<>();
        q.add(new int[] { i, j });
        visited[i][j] = true;

        System.out.println(q);
        while (!q.isEmpty()) {
            int x = q.peek()[0];
            int y = q.peek()[1];
            q.poll();
            for (int to = 0; to < 4; to++) {
                int ni = x + dx[to];
                int nj = y + dy[to];

                if (nj < 0 || ni < 0 || ni >= m || nj >= n)
                    continue;
                if (grid[ni][nj] == '0')
                    continue;
                if (visited[ni][nj])
                    continue;

                q.add(new int[] { ni, nj });
                visited[ni][nj] = true;
            }
        }

    }

    public static void main(String[] args) {

        Solution200 s = new Solution200();

        char[][] grid = {
                { '1', '1', '0', '0', '0' },
                { '1', '1', '0', '0', '0' },
                { '0', '0', '1', '0', '0' },
                { '0', '0', '0', '1', '1' },
        };

        int res = s.numIslands(grid);

        System.out.println(res);

        // Queue q = new ArrayDeque<>();
        // q.add(1);
        // q.add(2);
        // q.add(3);
        // System.out.println(q);
        // System.out.println(q.peek());
        // q.poll();
        // System.out.println(q);

    }

}

File: Search/src/main/python/bfs/solution200.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dx = [1, 0, 0, -1]
        dy = [0, 1, -1, 0]

        m = len(grid)
        n = len(grid[0])

        visited = [[False for _ in range(n)] for _ in range(m)]

        def bfs(i, j):
            queue = []
            queue.append((i, j))
            visited[i][j] = True
            while len(queue) != 0:
                # 出队
                i = queue[0][0]
                j = queue[0][1]
                queue.pop(0)
                for to in range(4):
                    ni = i + dx[to]
                    nj = j + dy[to]
                    if ni < 0 or nj < 0 or ni >= m or nj >= n:
                        continue
                    if visited[ni][nj]:
                        continue
                    if grid[ni][nj] == "0":
                        continue
                    queue.append((ni, nj))
                    visited[ni][nj] = True

        ans = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1" and not visited[i][j]:
                    ans += 1
                    bfs(i, j)

        return ans


s = Solution()
grid = [
    ["1", "1", "0", "0", "0"],
    ["1", "1", "0", "0", "0"],
    ["0", "0", "1", "0", "0"],
    ["0", "0", "0", "1", "1"],
]


res = s.numIslands(grid=grid)
print(res)
